editor/modes/visual-editor/block.js

```diff
@@ -251,68 +251,161 @@ class VisualEditorBlock extends Component {
-
-		if ( moveUp || moveDown ) {
-			const selection = window.getSelection();
-			const range = selection.rangeCount ? selection.getRangeAt( 0 ) : null;
-
-			// If there's no movement, so we're either at the end of start, or
-			// no text input at all.
-			if ( range !== this.lastRange ) {
+			if ( ! reverse && offset !== node.data.length ) {
+				return false;
+			}
+
+			while ( node !== rootNode ) {
+				const parentNode = node.parentNode;
+
+				if ( parentNode[ `${ order }Child` ] !== node ) {
+					return false;
+				}
+
+				node = parentNode;
+			}
+
+			return true;
+		};
+
+		const getNextEditor = ( { node, reverse } ) => {
+			const selector = '.editor-visual-editor [contenteditable="true"]';
+			const editableNodes = Array.from( document.querySelectorAll( selector ) );
+
+			if ( reverse ) {
+				editableNodes.reverse();
+			}
+
+			const index = editableNodes.indexOf( target );
+			const nextEditableNode = editableNodes[ index + 1 ];
+			const direction = reverse ? 'previous' : 'next';
+			const nextBlockNode = node[ `${ direction }Sibling` ];
+
+			if ( ! nextBlockNode ) {
 				return;
 			}
 
-			const focusableNodes = Array.from( document.querySelectorAll( selectors ) );
+			const editor = tinymce.get( nextEditableNode.id );
 
-			if ( moveUp ) {
-				focusableNodes.reverse();
+			if ( ! editor ) {
+				return;
 			}
 
-			const targetNode = focusableNodes
-				.slice( focusableNodes.indexOf( target ) )
-				.reduce( ( result, node ) => {
-					return result || ( node.contains( target ) ? null : node );
-				}, null );
+			if ( ! node.contains( nextEditableNode ) && ! nextBlockNode.contains( nextEditableNode ) ) {
+				return;
+			}
+
+			return editor;
+		};
+
+		if ( up || down || left || right ) {
+			const editor = target.id && tinymce.get( target.id );
+
+			if ( target === this.node ) {
+				const reverse = up || left;
+				const followingBlock = reverse ? previousBlock : nextBlock;
+
+				if ( followingBlock ) {
+					event.preventDefault();
+					onFocus( followingBlock.uid, { offset: reverse ? -1 : 0 } );
+				}
+			}
+
+			if ( editor ) {
+				const reverse = up || left;
+				const horizontal = left || right;
+				const followingBlock = reverse ? previousBlock : nextBlock;
+
+				const rect = horizontal ?
+					isHorizontalEdge( { editor, reverse } ) :
+					isVerticalEdge( { editor, reverse } );
+
+				if ( ! rect ) {
+					return;
+				}
+
+				event.preventDefault();
+
+				const followingEditor = getNextEditor( { node: this.node, reverse } );
 
-			if ( targetNode ) {
-				targetNode.focus();
+				if ( ! followingEditor ) {
+					if ( followingBlock ) {
+						onFocus( followingBlock.uid, { offset: reverse ? -1 : 0 } );
+					}
+
+					return;
+				}
+
+				followingEditor.focus();
+
+				if ( horizontal ) {
+					if ( reverse ) {
+						followingEditor.selection.select( followingEditor.getBody(), true );
+						followingEditor.selection.collapse( false );
+					}
+				} else {
+					window.setTimeout( () => {
```

Comment:
@iseulde I created a fiddle here seems to work if you first set the selection then you move the focus. http://fiddle.tinymce.com/qXfaab/1 This seems to be a browser thing rather than a tinymce thing since I tired manually setting focus as well and that didn't work. Tested it on IE, FF, Chrome, Edge, Safari. I needed to set it twice in safari though for some odd reason.